### PR TITLE
Divide `NO_ITERATORS` into two counters `NO_ITERATOR_CREATED` and `NO_ITERATOR_DELETE`

### LINKED ISSUE #3013: RecordTick in DbIter __dtor potentially wrong
> Note: Please use Issues only for bug reports. For questions, discussions, feature requests, etc. post to dev group: https://www.facebook.com/groups/rocksdb.dev

https://github.com/facebook/rocksdb/blob/5.11.fb/db/db_iter.cc#L153 potentially make the stats wrong.

The argument of -1 for uint64_t results in a max value for the type. 

### Expected behavior

The number of ticks decreased

### Actual behavior

Depending on the impl we have a potential overflow.

### Steps to reproduce the behavior

Code review

### PR DESCRIPTION
**Pull request by solicomo**
Summary:

Currently, `Statistics` can record tick by `recordTick()` whose second parameter is an `uint64_t`.
That means tick can only increase.
If we want to reduce tick, we have to work around like `RecordTick(statistics_, NO_ITERATORS, uint64_t(-1));`.
That's kind of a hack.

So, this PR divide `NO_ITERATORS` into two counters `NO_ITERATOR_CREATED` and `NO_ITERATOR_DELETE`, making the counters increase only.

Fixes #3013 .

### CODE REVIEW

File: monitoring/statistics.cc [hunk 1]
```diff
@@ -121,6 +121,18 @@ void StatisticsImpl::recordTick(uint32_t tickerType, uint64_t count) {
   }
 }
 
+void StatisticsImpl::reduceTick(uint32_t tickerType, uint64_t count) {
+  assert(
+    enable_internal_stats_ ?
+      tickerType < INTERNAL_TICKER_ENUM_MAX :
+      tickerType < TICKER_ENUM_MAX);
+  per_core_stats_.Access()->tickers_[tickerType].fetch_sub(
```
Comment (yiwu-arbug) [inline monitoring/statistics.cc]: assert the return value not smaller than `count`?

Comment (solicomo) [reply monitoring/statistics.cc]: Good idea.

**Comment by yiwu-arbug** [comment]
DBTest.GroupCommitTest is failing in AppVeyor for failing the new assert. The test was flaky before and I'm guessing the new assert catch something wrong with the test. But I'm not familiar with that test.

**Comment by solicomo** [comment]
>           DBTest.GroupCommitTest is failing in AppVeyor for failing the new assert. The test was flaky before and I'm guessing the new assert catch something wrong with the test. But I'm not familiar with that test.

Yes. I noticed that.
I am working on it.
Hope I can fix the flaky test failure by the way.

**Comment by yiwu-arbug** [comment]
`DBIteratorTestInstance/DBIteratorTest.IteratorWithLocalStatistics/0` is failing in our internal CI,  on the same assert.

**Comment by solicomo** [comment]
> `DBIteratorTestInstance/DBIteratorTest.IteratorWithLocalStatistics/0` is failing in our internal CI, on the same assert.

Should I remove the assertion for now, and add it back after we figure out the reason of test failures?

**Comment by yiwu-arbug** [comment]
@solicomo I'll take a look by the end of the week.

**Comment by solicomo** [comment]
> @solicomo I'll take a look by the end of the week.

I think I found the reason.

`StatisticsImpl` is using `CoreLocalArray` to store the statistics data.
But `DBIter::DBIter()` and `DBIter::~DBIter()` are NOT always running on the same core.

Now I understand why @siying said "statistics is not the place to keep current status".

Hi @siying , do you have any idea now?

**Comment by siying** [comment]
@solicomo I'm still not fan of using statistics as current status. I hope counters are only increasing.

**Comment by yiwu-arbug** [comment]
@solicomo well, in that case, the assert I asked to add is incorrect.

**Comment by solicomo** [comment]
> @solicomo I'm still not fan of using statistics as current status. I hope counters are only increasing.

"counters are only increasing" makes sense to me.

So, how about not reducing `NO_ITERATORS` in destructor and only increasing it in constructor?
Then `ReduceTick()` is not necessary.

The solution turns out to be extremely simple - just remove [this line](https://github.com/facebook/rocksdb/blob/master/db/db_iter.cc#L163):

`RecordTick(statistics_, NO_ITERATORS, uint64_t(-1));`


@yiwu-arbug , @siying  How do you think?

**Comment by yiwu-arbug** [comment]
It is up to @siying. I'm fine with either.

**Comment by yiwu-arbug** [comment]
I'm fine with either way. @siying What's your thinking? :)

**Comment by siying** [comment]
Yes I'm fine. It's basically what we do for BLOCK_CACHE_INDEX_BYTES_INSERT and BLOCK_CACHE_INDEX_BYTES_EVICT. People take a difference of the two and get the value. We can follow the same pattern. Divide the counter into two.

**Comment by solicomo** [comment]
> Yes I'm fine. It's basically what we do for BLOCK_CACHE_INDEX_BYTES_INSERT and BLOCK_CACHE_INDEX_BYTES_EVICT. People take a difference of the two and get the value. We can follow the same pattern. Divide the counter into two.

That's a good idea.
I am working on it.

By the way, should we keep `ReduceTick()`?

**Comment by yiwu-arbug** [comment]
@solicomo let's discard `ReduceTick()` then.

**Comment by yiwu-arbug** [comment]
since we don't want to encourage people to reduce tick, and we don't have a proper implementation to support cross-core reduce tick.

**Comment by solicomo** [comment]
rebase and give AppVeyor another try.

**Comment by solicomo** [comment]
The test failure might be not related to my fix.
Other PRs also encounter this failure.

By the way, I changed the PR title.

**Comment by yiwu-arbug** [comment]
@solicomo Would you also update HISTORY.md about the change, under "Unreleased - Public API" section? Thanks.

**Comment by solicomo** [comment]
> @solicomo Would you also update HISTORY.md about the change, under "Unreleased - Public API" section? Thanks.

HISTORY.md is updated. 
And all checks have passed this time.

**Comment by yiwu-arbug** [comment]
@solicomo Thanks much! Merging.

**Comment by siying** [comment]
Hmm. This is not merged. Let me try again.

**Comment by miasantreble** [comment]
@siying landing failed because of test failure: 
> DBCompactionTestWithParam/DBCompactionTestWithParam.IntraL0CompactionDoesNotObsoleteDeletions/3

**Comment by solicomo** [comment]
Rebase and trigger another round of testing.

**Comment by solicomo** [comment]
> @siying landing failed because of test failure:
> 
> > DBCompactionTestWithParam/DBCompactionTestWithParam.IntraL0CompactionDoesNotObsoleteDeletions/3

Hi, @miasantreble , I rebased my commits and triggered testing again.
Now all the tests look good.

**Comment by sagar0** [comment]
Resolved HISTORY.md merge conflicts.
